libs/ml/include/ml/ops/matrix_multiply.hpp

```diff
@@ -36,36 +37,215 @@ class MatrixMultiply : public fetch::ml::BatchOps<T>
   MatrixMultiply()  = default;
   ~MatrixMultiply() = default;
 
-  void Forward(VecTensorType const &inputs, ArrayType &output)
+  void                   Forward(VecTensorType const &inputs, ArrayType &output);
+  std::vector<ArrayType> Backward(VecTensorType const &inputs, ArrayType const &error_signal);
+  std::vector<SizeType>  ComputeOutputShape(VecTensorType const &inputs) const;
+
+  static constexpr char const *DESCRIPTOR = "MatrixMultiply";
+};
+
+template <class T>
+void MatrixMultiply<T>::Forward(VecTensorType const &inputs, ArrayType &output)
+{
+  ASSERT(inputs.size() == 2);
+  ASSERT(output.shape() == ComputeOutputShape(inputs));
+
+  // Normal MatMul 2D @ 2D
+  if (inputs.at(0).get().shape().size() == 2 && inputs.at(1).get().shape().size() == 2)
   {
-    ASSERT(inputs.size() == 2);
-    ASSERT(inputs.at(0).get().shape().size() == 2);
-    ASSERT(inputs.at(1).get().shape().size() == 2);
-    ASSERT(output.shape() == ComputeOutputShape(inputs));
+    fetch::math::Dot(inputs.at(0).get(), inputs.at(1).get(), output);
+  }
+  // Batchwise 3D @ 3D or broadcast matmul 2D @ 3D, 3D @ 2D
+  else
+  {
+    ASSERT((inputs.at(0).get().shape().size() == 3 || inputs.at(0).get().shape().size() == 2) &&
+           (inputs.at(1).get().shape().size() == 3 || inputs.at(1).get().shape().size() == 2));
+
+    // Get batch size
+    SizeType batch_size;
+    if (inputs.at(0).get().shape().size() == 3)
+    {
+      batch_size = inputs.at(0).get().shape().at(2);
+    }
+    else
+    {
+      batch_size = inputs.at(1).get().shape().at(2);
+    }
 
-    fetch::math::Dot(inputs[0].get(), inputs[1].get(), output);
+    // Iterate over batch
+    for (SizeType i{0}; i < batch_size; i++)
+    {
+
+      auto      output_slice        = output.Slice(i, 2);
```

Comment:
Can I avoid that?
      // Copy data to original array
      output_slice.Assign(output_slice_tensor);

Reply:
Are you sure that it will work?
```: error: conversion from ‘fetch::math::Tensor<double>::TensorSlice’ to non-scalar type ‘fetch::ml::ops::MatrixMultiply<fetch::math::Tensor<double> >::ArrayType {aka fetch::math::Tensor<double>}’ requested
       ArrayType      output_slice_tensor        = output.Slice(i, 2);```